c/misra/src/rules/RULE-21-13/CtypeFunctionArgNotUnsignedCharOrEof.ql

```diff
@@ -0,0 +1,50 @@
+/**
+ * @id c/misra/ctype-function-arg-not-unsigned-char-or-eof
+ * @name RULE-21-13: <ctype.h> function arguments shall be represented as unsigned char
+ * @description Passing arguments to <ctype.h> functions outside the range of unsigned char or EOF
+ *              causes undefined behavior.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity error
+ * @tags external/misra/id/rule-21-13
+ *       external/misra/obligation/mandatory
+ */
+
+import cpp
+import codingstandards.c.misra
+import codingstandards.cpp.ReadErrorsAndEOF
+import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
+import semmle.code.cpp.rangeanalysis.RangeAnalysisUtils
+
+class CtypeFunction extends Function {
+  CtypeFunction() { this.getADeclaration().getAFile().(HeaderFile).getShortName() = "_ctype" }
+}
+
+predicate unsignedCharRange(int lower, int upper, EOFInvocation eof) {
+  exists(UnsignedCharType unsignedChar |
+    lower = typeLowerBound(unsignedChar) and
+    upper = upperBound(eof.getExpr()) and
+    typeLowerBound(unsignedChar) <= lowerBound(eof.getExpr()) and
+    upperBound(eof.getExpr()) <= typeUpperBound(unsignedChar)
+  )
+}
+
+predicate isEquivToEOF(Expr expr) {
+  exists(EOFInvocation eof | DataFlow::localFlow(DataFlow::exprNode(eof.getExpr()), DataFlow::exprNode(expr)))
+}
+
+from FunctionCall ctypeCall
+where
+  not isExcluded(ctypeCall,
+    StandardLibraryFunctionTypesPackage::ctypeFunctionArgNotUnsignedCharOrEofQuery()) and
+  exists(CtypeFunction ctype, UnsignedCharType unsignedChar |
+    ctypeCall = ctype.getACallToThisFunction()
+  |
```

Comment:
This, very strangely, captures `c2` from `isDigit(c2)` in the `test.c`; the range analysis computes the lower bound and upper bound of `c2` as 0 and 255, respectively. However, intuitively it should not because of EOF is defined as `#define EOF -1`.

Therefore, I have added an extra def-use check below to see if the `EOF` directly flowed to an argument.